### PR TITLE
[SYCL][ESIMD] Add ESIMD-specific IR verification pass

### PR DESCRIPTION
**Pull request by sndmitriev**
Signed-off-by: Sergey Dmitriev <serguei.n.dmitriev@intel.com>

### CODE REVIEW

File: llvm/include/llvm/IR/ESIMDVerifier.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
+//===--------- ESIMDVerifier.h - ESIMD-specific IR verification -----------===//
```
Comment (kbobrovs) [inline llvm/include/llvm/IR/ESIMDVerifier.h:1]: For easier potential upstream, I'd suggest to keep it next to other SYCL/ESIMD-specific passes - llvm\lib\SYCLLowerIR, llvm\include\llvm\SYCLLowerIR.

File: llvm/lib/IR/ESIMDVerifier.cpp [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Regex.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "esimd-verifier"
+
+// A list of unsupported functions in ESIMD context.
+static const char *IllegalFunctions[] = {
+    "^cl::sycl::multi_ptr<.+> cl::sycl::accessor<.+>::get_pointer<.+>\\(\\) "
```
Comment (kbobrovs) [inline llvm/lib/IR/ESIMDVerifier.cpp:30]: 'cl' is deprecated and maybe removed soon, suggest to skip it

File: sycl/test/esimd/vadd.cpp [hunk 3]
```diff
@@ -8,6 +8,10 @@
 // RUN: %clangxx -I %sycl_include %s -o %t.out -fsycl \
 // RUN: -fsycl-instrument-device-code
 
+// Adding xfail because cl::sycl::accessor<...>::get_pointer<...>() is not
+// supported in ESIMD context.
+// XFAIL: *
```
Comment (kbobrovs) [inline sycl/test/esimd/vadd.cpp:13]: could you please fix it by converting
```
            simd<int, VL> va = gather<int, VL>(pA + i * VL, offsets);
            simd<int, VL> vb = block_load<int, VL>(pB + i * VL);
            simd<int, VL> vc = va + vb;
            block_store<int, VL>(pC + i * VL, vc);
```
to accessor-based calls or converting
```
      auto accA = bufA.get_access<cl::sycl::access::mode::read>(cgh);
      auto accB = bufB.get_access<cl::sycl::access::mode::read>(cgh);
      auto accC = bufC.get_access<cl::sycl::access::mode::write>(cgh);
```
to USM pointers?

File: llvm/lib/Passes/PassBuilderPipelines.cpp [hunk 4]
```diff
@@ -735,6 +739,9 @@ PassBuilder::buildModuleSimplificationPipeline(OptimizationLevel Level,
                                                ThinOrFullLTOPhase Phase) {
   ModulePassManager MPM;
 
+  if (SYCLOptimizationMode)
```
Comment (kbobrovs) [inline llvm/lib/Passes/PassBuilderPipelines.cpp:742]: I'd suggest to add this pass to clang\lib\CodeGen\BackendUtil.cpp, similar to other SYCL-specific passes like `createSYCLLowerWGLocalMemoryLegacyPass`

File: llvm/lib/Transforms/IPO/PassManagerBuilder.cpp [hunk 5]
```diff
@@ -675,6 +675,9 @@ void PassManagerBuilder::populateModulePassManager(
 
   MPM.add(createAnnotation2MetadataLegacyPass());
 
+  if (SYCLOptimizationMode)
```
Comment (kbobrovs) [inline llvm/lib/Transforms/IPO/PassManagerBuilder.cpp:678]: ... if added to BackendUtil.cpp, changes here should not be needed, right? Also, the verifier should work at all opt levels and be enabled by default, I believe.

File: llvm/include/llvm/IR/ESIMDVerifier.h (hunk 1 above)
Comment (sndmitriev) [reply llvm/include/llvm/IR/ESIMDVerifier.h]: Moved to SYCLLowerIR

File: llvm/lib/Passes/PassBuilderPipelines.cpp (hunk 4 above)
Comment (sndmitriev) [reply llvm/lib/Passes/PassBuilderPipelines.cpp]: Done

File: sycl/test/esimd/vadd.cpp (hunk 3 above)
Comment (sndmitriev) [reply sycl/test/esimd/vadd.cpp]: Converted to USM pointers.

File: clang/lib/CodeGen/BackendUtil.cpp [hunk 6]
```diff
@@ -849,6 +850,9 @@ void EmitAssemblyHelper::CreatePasses(legacy::PassManager &MPM,
     FPM.add(createVerifierPass());
 
   // Set up the per-module pass manager.
+  if (CodeGenOpts.VerifyModule && LangOpts.SYCLIsDevice)
```
Comment (kbobrovs) [inline clang/lib/CodeGen/BackendUtil.cpp:853]: Let's add this pass unconditionally, as it is not really a verification, it is semantic analysis normally done by FE

**Review by kbobrovs — APPROVED** [review]
thanks!

**Review by bader — APPROVED** [review]

**Review by smanna12 — APPROVED** [review]
FE changes look good to me.

**Pull request merged**